src/hypofuzz/provider.py

```diff
@@ -26,29 +80,409 @@ def fresh_choice(
+            # if no worker has ever worked on this test before, save an initial
+            # Phase.GENERATE report, so the graph displays a point at (0, 0).
+            #
+            # This is a hack. We should move this to the dashboard plotting code
+            # instead, by adding a virtual (0, 0) point.
+            self.phase = Phase.GENERATE
+            self._save_report(self._report)
+            self.phase = None
+
+        self._started = True
+
+    def _enqueue(
+        self,
+        priority: ReplayPriority,
+        choices: ChoicesT,
+        *,
+        extra_data: Optional[Any] = None,
+    ) -> None:
+        self._replay_queue.add((priority, choices, extra_data))
+
+    def on_event(self, event: DatabaseEvent) -> None:
+        # Some worker has found a new covering corpus element. Replay it in
+        # this worker.
+        if event.type == "save":
+            if event.key is DatabaseEventKey.CORPUS:
+                # the worker which saved this choice sequence might have been *us*,
+                # or we might simply already have this choice sequence in our
+                # corpus.
+                if (
+                    self.corpus is not None
+                    and Choices(event.value) in self.corpus.corpus
+                ):
+                    return
+                self._enqueue(ReplayPriority.COVERING, event.value)
+            # (should we also replay failures found by other workers? we may not
+            # want to stop early, since we could still find a different failure.
+            # but I guess the same logic would apply to the worker which found the
+            # failure..)
+
+    def _start_phase(self, phase: Phase) -> None:
+        if phase is self.phase:
+            return
+        if self.phase is not None:
+            # don't save a report the very first time we start a phase
+            self._save_report(self._report)
+        self.phase = phase
+
+    def _should_save_timed_report(self) -> bool:
+        # linear interpolation from 1 report/s at the start to 1 report/60s after
+        # 5 minutes have passed
+        increment = lerp(1, 60, min(self._last_saved_report_at / 60 * 5, 1))
+        # A "timed report" is one that we expect to be discarded from the database
+        # on the next saved report, but which serves as an incremental progress
+        # marker for the dashboard.
+        return self.elapsed_time > self._last_saved_report_at + increment
+
+    def _save_report(self, report: Report) -> None:
+        assert self.database_key is not None
+        assert self.corpus is not None
+        assert self.db is not None
+
+        self.db.save_report(self.database_key, report)
+        self._last_saved_report_at = self.elapsed_time
+
+        # Having written the latest report, we can keep the database small
+        # by dropping the previous report unless it differs from the latest in
+        # an important way.
+        if self._last_report and not (
```

Comment:
for a followup PR: let's instead track `self._last_timed_report`, or another mechanism to avoid the status quo where we keep the last timed report made before each event report.